include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
+
+        retval.TdotL = tsL.x;
+        retval.BdotL = tsL.y;
+        retval.NdotL = tsL.z;
+        retval.NdotL2 = retval.NdotL*retval.NdotL;
+
+        return retval;
+    }
+    static this_t create(NBL_CONST_REF_ARG(RayDirInfo) L, const scalar_type VdotL, NBL_CONST_REF_ARG(vector3_type) N)
+    {
+        this_t retval;
+
+        retval.L = L;
+        retval.VdotL = VdotL;
+
+        retval.TdotL = nbl::hlsl::numeric_limits<scalar_type>::quiet_NaN;
+        retval.BdotL = nbl::hlsl::numeric_limits<scalar_type>::quiet_NaN;
+        retval.NdotL = nbl::hlsl::dot<vector3_type>(N,L.getDirection());
+        retval.NdotL2 = retval.NdotL * retval.NdotL;
+
+        return retval;
+    }
+    static this_t create(NBL_CONST_REF_ARG(RayDirInfo) L, const scalar_type VdotL, NBL_CONST_REF_ARG(vector3_type) T, NBL_CONST_REF_ARG(vector3_type) B, NBL_CONST_REF_ARG(vector3_type) N)
+    {
+        this_t retval = create(L,VdotL,N);
+
+        retval.TdotL = nbl::hlsl::dot<vector3_type>(T,L.getDirection());
+        retval.BdotL = nbl::hlsl::dot<vector3_type>(B,L.getDirection());
+
+        return retval;
+    }
+    // overloads for surface_interactions, NOTE: temporarily commented out due to dxc bug https://github.com/microsoft/DirectXShaderCompiler/issues/7154
+    // template<class ObserverRayDirInfo>
+    // static this_t create(NBL_CONST_REF_ARG(vector3_type) L, NBL_CONST_REF_ARG(surface_interactions::SIsotropic<ObserverRayDirInfo>) interaction)
+    // {
+    //     const vector3_type V = interaction.V.getDirection();
+    //     const scalar_type VdotL = nbl::hlsl::dot<vector3_type>(V,L);
+    //     return create(L, VdotL, interaction.N);
+    // }
+    // template<class ObserverRayDirInfo>
+    // static this_t create(NBL_CONST_REF_ARG(vector3_type) L, NBL_CONST_REF_ARG(surface_interactions::SAnisotropic<ObserverRayDirInfo>) interaction)
+    // {
+    //     const vector3_type V = interaction.V.getDirection();
+    //     const scalar_type VdotL = nbl::hlsl::dot<vector3_type>(V,L);
+    //     return create(L,VdotL,interaction.T,interaction.B,interaction.N);
+    // }
+    //
+    vector3_type getTangentSpaceL() NBL_CONST_MEMBER_FUNC
+    {
+        return vector3_type(TdotL, BdotL, NdotL);
+    }
+
+    RayDirInfo getL() NBL_CONST_MEMBER_FUNC { return L; }
+    scalar_type getVdotL() NBL_CONST_MEMBER_FUNC { return VdotL; }
+    scalar_type getTdotL() NBL_CONST_MEMBER_FUNC { return TdotL; }
+    scalar_type getBdotL() NBL_CONST_MEMBER_FUNC { return BdotL; }
+    scalar_type getNdotL() NBL_CONST_MEMBER_FUNC { return NdotL; }
+    scalar_type getNdotL2() NBL_CONST_MEMBER_FUNC { return NdotL2; }
+
+
+    RayDirInfo L;
+    scalar_type VdotL;
+
+    scalar_type TdotL;
+    scalar_type BdotL;
+    scalar_type NdotL;
+    scalar_type NdotL2;
 };
 
-struct AnisotropicMicrofacetCache : IsotropicMicrofacetCache
+
+// TODO: figure out the commented constraints, templated RayDirInfo not really working for some reason
```

Comment:
probably because your code is written to require a sample and not a ray dir info for the other direction